docs(uno): Simple photo viewer tutorial (#1)
lukeblevins/windows-dev-docs: hub/apps/get-started/uno-simple-photo-viewer.md

Reviewer: These phrases:
"Learn how to reach users on other platforms like Web, iOS, Android, and Linux ..." (line3)
"... app ported to platforms like the web, mobile, and desktop." (line 14)

Is it platforms **like** these, or these actual platforms?
Why is "web" inconsistently capitalized?
Why the change between naming operating systems and then talking about them in the abstract (mobile & desktop)?

Author: Thank you Matt.

I've addressed this feedback by using clearer language when describing use across the supported platforms. The change also refers to what we support more consistently without enumerating every OS each time. We also now lowercase references to "the web".

---
title: Tutorial--Make a simple photo viewer that targets multiple platforms

> description: Learn how to reach users on other platforms like Web, iOS, and Android with minimal changes to the C#/WinUI 3 simple photo viewer built in the previous tutorial. We'll use Uno Platform to create a new multi-platform app, which we can move code from the existing desktop project to.

Reviewer 2: Is there a reason Linux is omitted? Ideally we have it in

> description: Learn how to reach users on other platforms like Web, iOS, and Android with minimal changes to the C#/WinUI 3 simple photo viewer built in the previous tutorial. We'll use Uno Platform to create a new multi-platform app, which we can move code from the existing desktop project to.

Reviewer 2: Please hyperlink Uno Platform website

ms.topic: article
ms.date: 05/21/2023

> keywords: Windows, App, SDK, WinUI 3, WinUI, photo, viewer, Windows 11, Windows 10, XAML, C#, uno platform, uno

Reviewer 2: should other platforms be added?

Author: @sasakrsmanovic This is something I'll check with Alvin about

ms.author: aashcraft
author: alvinashcraft
ms.localizationpriority: medium
---

# Tutorial: Make a simple photo viewer that targets multiple platforms

> After you've [created](/hub/apps/get-started/simple-photo-viewer-winui3.md) a starter simple photo viewer WinUI 3 app, you might be wondering how to reach more users without having to rewrite your app. This tutorial will use [Uno Platform](https://platform.uno/) to expand the reach of your existing C# WinUI 3 application enabling reuse of the business logic and UI layer across native mobile, web, and desktop. With only minimal changes to the simple photo viewer app, we'll be able to run a pixel-perfect copy of the app ported to platforms like Web, iOS, Android, and macOS.

Reviewer 2: Linux remark again

> After you've [created](/hub/apps/get-started/simple-photo-viewer-winui3.md) a starter simple photo viewer WinUI 3 app, you might be wondering how to reach more users without having to rewrite your app. This tutorial will use [Uno Platform](https://platform.uno/) to expand the reach of your existing C# WinUI 3 application enabling reuse of the business logic and UI layer across native mobile, web, and desktop. With only minimal changes to the simple photo viewer app, we'll be able to run a pixel-perfect copy of the app ported to platforms like Web, iOS, Android, and macOS.

Reviewer 3: ```suggestion
After you've [created](/hub/apps/get-started/simple-photo-viewer-winui3.md) a starter simple photo viewer WinUI 3 app, you might be wondering how to reach more users without having to rewrite your app. This tutorial will use [Uno Platform](https://platform.uno/) to expand the reach of your existing C# WinUI 3 application enabling reuse of the business logic and UI layer across native mobile, web, and desktop. With only minimal changes to the simple photo viewer app, we'll be able to run a pixel-perfect copy of the app ported to platforms like the Web, iOS, Android, and macOS.
```

:::image type="content" source="../images/vs2022-project-reload.png" alt-text="Visual Studio banner offering to reload your projects to complete changes":::

## Preparing your app

You should see the following default file structure in your **Solution Explorer**:

:::image type="content" source="images/hello-world/uno-file-structure.png" alt-text="Default file structure":::

Now that you've generated the functional starting point of your multi-platform WinUI application, you can copy code into it from the desktop project. Because Uno Platform allows you to use the XAML flavor you're already familiar with, you can copy over the same code you created in the [previous tutorial](/hub/apps/get-started/simple-photo-viewer-winui3.md). 

To do so, go back to the **SimplePhotos** project from the previous tutorial. In the **Solution Explorer**, find the file named `MainWindow.xaml` and open it. Observe that the contents of the view are defined within a `Window` element rather than a `Page`. This is because the desktop project is a WinUI 3 application, which can use `Window` elements to define the contents of the view.

```xml

> <Window x:Class="SimplePhotos.MainWindow"
>     xmlns="http://schemas.microsoft.com/winfx/2006/xaml/presentation"
> 	xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"

Reviewer 3: ```suggestion
<Window x:Class="SimplePhotos.MainWindow"
	xmlns="http://schemas.microsoft.com/winfx/2006/xaml/presentation"
	xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
```

	xmlns:local="using:SimplePhotos"
	xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
	xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
	mc:Ignorable="d">

    <Grid>
        <Grid.Resources>

>             <DataTemplate x:Key="ImageGridView_ItemTemplate"
>                         x:DataType="local:ImageFileInfo">

Reviewer 3: ```suggestion
            <DataTemplate x:Key="ImageGridView_ItemTemplate"
                          x:DataType="local:ImageFileInfo">
```

>                 <Grid Height="300"
>                     Width="300"
>                     Margin="8">

Reviewer 3: ```suggestion
                <Grid Height="300"
                      Width="300"
                      Margin="8">
```

                    <Grid.RowDefinitions>
                        <RowDefinition />
                        <RowDefinition Height="Auto" />
                    </Grid.RowDefinitions>

>                     <Image x:Name="ItemImage"
>                         Source="Assets/StoreLogo.png"
>                         Stretch="Uniform" />

Reviewer 3: ```suggestion
                    <Image x:Name="ItemImage"
                           Source="Assets/StoreLogo.png"
                           Stretch="Uniform" />
```

                    <StackPanel Orientation="Vertical"
                                Grid.Row="1">

>                         <TextBlock Text="{x:Bind ImageTitle}"
>                                 HorizontalAlignment="Center"
>                                 Style="{StaticResource SubtitleTextBlockStyle}" />

Reviewer 3: ```suggestion
                        <TextBlock Text="{x:Bind ImageTitle}"
                                   HorizontalAlignment="Center"
                                   Style="{StaticResource SubtitleTextBlockStyle}" />
```

                        <StackPanel Orientation="Horizontal"
                                    HorizontalAlignment="Center">

>                             <TextBlock Text="{x:Bind ImageFileType}"
>                                     HorizontalAlignment="Center"
>                                     Style="{StaticResource CaptionTextBlockStyle}" />

Reviewer 3: ```suggestion
                            <TextBlock Text="{x:Bind ImageFileType}"
                                       HorizontalAlignment="Center"
                                       Style="{StaticResource CaptionTextBlockStyle}" />
```

>                             <TextBlock Text="{x:Bind ImageDimensions}"
>                                     HorizontalAlignment="Center"
>                                     Style="{StaticResource CaptionTextBlockStyle}"
>                                     Margin="8,0,0,0" />

Reviewer 3: ```suggestion
                            <TextBlock Text="{x:Bind ImageDimensions}"
                                       HorizontalAlignment="Center"
                                       Style="{StaticResource CaptionTextBlockStyle}"
                                       Margin="8,0,0,0" />
```

                        </StackPanel>

>                         <RatingControl Value="{x:Bind ImageRating}" IsReadOnly="True"/>

Reviewer 3: ```suggestion
                        <RatingControl Value="{x:Bind ImageRating}" 
                                       IsReadOnly="True"/>
```

                    </StackPanel>
                </Grid>
            </DataTemplate>

>             <Style x:Key="ImageGridView_ItemContainerStyle"
>                 TargetType="GridViewItem">

Reviewer 3: ```suggestion
            <Style x:Key="ImageGridView_ItemContainerStyle"
                   TargetType="GridViewItem">
```

                <Setter Property="Background" Value="Gray"/>
                <Setter Property="Margin" Value="8"/>
            </Style>

            <ItemsPanelTemplate x:Key="ImageGridView_ItemsPanelTemplate">

>                     <ItemsWrapGrid Orientation="Horizontal"
>                                 HorizontalAlignment="Center"/>

Reviewer 3: ```suggestion
                    <ItemsWrapGrid Orientation="Horizontal"
                                   HorizontalAlignment="Center"/>
```

                </ItemsPanelTemplate>
        </Grid.Resources>

>         <GridView x:Name="ImageGridView"
>                 ItemsSource="{x:Bind Images}"
>                 ItemTemplate="{StaticResource ImageGridView_ItemTemplate}"
>                 ItemContainerStyle="{StaticResource ImageGridView_ItemContainerStyle}"
>                 ItemsPanel="{StaticResource ImageGridView_ItemsPanelTemplate}"
>                 ContainerContentChanging="ImageGridView_ContainerContentChanging">
>         </GridView>

Reviewer 3: ```suggestion
        <GridView x:Name="ImageGridView"
                  ItemsSource="{x:Bind Images}"
                  ItemTemplate="{StaticResource ImageGridView_ItemTemplate}"
                  ItemContainerStyle="{StaticResource ImageGridView_ItemContainerStyle}"
                  ItemsPanel="{StaticResource ImageGridView_ItemsPanelTemplate}"
                  ContainerContentChanging="ImageGridView_ContainerContentChanging" />
```

                        </StackPanel>

                        <RatingControl Value="{x:Bind ImageRating}" IsReadOnly="True"/>
                    </StackPanel>
                </Grid>
            </DataTemplate>

            <Style x:Key="ImageGridView_ItemContainerStyle"
                TargetType="GridViewItem">
                <Setter Property="Background" Value="Gray"/>
                <Setter Property="Margin" Value="8"/>
            </Style>

            <ItemsPanelTemplate x:Key="ImageGridView_ItemsPanelTemplate">
                    <ItemsWrapGrid Orientation="Horizontal"
                                HorizontalAlignment="Center"/>
                </ItemsPanelTemplate>
        </Grid.Resources>

        <GridView x:Name="ImageGridView"
                ItemsSource="{x:Bind Images}"
                ItemTemplate="{StaticResource ImageGridView_ItemTemplate}"
                ItemContainerStyle="{StaticResource ImageGridView_ItemContainerStyle}"
                ItemsPanel="{StaticResource ImageGridView_ItemsPanelTemplate}"
                ContainerContentChanging="ImageGridView_ContainerContentChanging">
        </GridView>
    </Grid>

> </Page>

Reviewer 3: Same here, formatting of the XAML

                _ = ImageProperties.SavePropertiesAsync();
                OnPropertyChanged();
            }
        }
    }

    public int ImageRating
    {
        get => (int)ImageProperties.Rating;
        set
        {
            if (ImageProperties.Rating != value)
            {
                ImageProperties.Rating = (uint)value;
                _ = ImageProperties.SavePropertiesAsync();
                OnPropertyChanged();
            }
        }
    }

    public event PropertyChangedEventHandler PropertyChanged;

    protected void OnPropertyChanged([CallerMemberName] string propertyName = null) =>
        PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(propertyName));
}
```

This class will serve as a model to represent the image files in the `GridView`. In the next sections, we will make a set of changes to these copied files to make them compatible in a multi-platform context.

### Add image assets to the project

In the `SimplePhotosUno` project, create a new folder named `Assets` and copy the JPEG image files to a `Samples` subfolder. The `Assets` folder structure should now look like this:

[Insert image here]

> ### Making the code multi-platform capable
> 
> In the desktop project from the previous tutorial, `MainPage.xaml.cs` file contains a `GetItemsAsync` method that enumerates items from a `StorageFolder` representing the installed package location. Because that location is not available on certain platforms such as WebAssembly, we will need to make changes to this method to make it compatible with all platforms. We will accordingly make some changes to the `ImageFileInfo` class to ensure compatibility.

Reviewer 3: ```suggestion
### Making the code multi-platform capable

In the desktop project from the previous tutorial, the `MainPage.xaml.cs` file contains a `GetItemsAsync` method that enumerates items from a `StorageFolder` representing the installed package location. Because that location is not available on certain platforms such as WebAssembly, we will need to make changes to this method to make it compatible with all platforms. We will accordingly make some changes to the `ImageFileInfo` class to ensure compatibility.
```

        string type,
        Uri uri)
    {
        ImageName = name;
        ImageFileType = type;
        ImageFile = imageFile;
        ImageUri = uri;
    }

    public StorageFile ImageFile { get; }

    public Uri ImageUri { get; }

    public string ImageName { get; }

    public string ImageFileType { get; }

    public event PropertyChangedEventHandler? PropertyChanged;

    protected void OnPropertyChanged([CallerMemberName] string? propertyName = null) =>
        PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(propertyName));
}
```

The `ImageFileInfo` class now contains a `Uri` property that will be used to load the image file. Both of the methods pertaining to getting an image source or thumbnail manually are now unneeded and can be removed. We have also omitted the `ImageDimensions` property because the `ImageProperties` type is not supported on all platforms. This `ImageFileInfo` class will be used to represent the image files in the `GridView`. We will now make changes to the `MainPage.xaml` file to accommodate these changes.

### Displaying images in GridView

In `MainPage.xaml`, replace the GridView element with the following code:

```xml

> <GridView x:Name="ImageGridView"
>         ItemsSource="{x:Bind Images, Mode=OneWay}"
>         ItemContainerStyle="{StaticResource ImageGridView_ItemContainerStyle}"
>         ItemTemplate="{StaticResource ImageGridView_ItemTemplate}"
>         ItemsPanel="{StaticResource ImageGridView_ItemsPanelTemplate}"/>

Reviewer 3: ```suggestion
<GridView x:Name="ImageGridView"
          ItemsSource="{x:Bind Images, Mode=OneWay}"
          ItemContainerStyle="{StaticResource ImageGridView_ItemContainerStyle}"
          ItemTemplate="{StaticResource ImageGridView_ItemTemplate}"
          ItemsPanel="{StaticResource ImageGridView_ItemsPanelTemplate}"/>
```

```

> Finally, modify the `ImageGridView_ItemTemplate` resource to use the `ImageUri` property of the `ImageFileInfo` class. We also removed elements which use the `ImageDimensions` property:

Reviewer 3: ```suggestion
Finally, modify the `ImageGridView_ItemTemplate` resource to use the `ImageUri` property of the `ImageFileInfo` class. We also removed elements that use the `ImageDimensions` property:
```

```xml
<DataTemplate x:Key="ImageGridView_ItemTemplate"
        x:DataType="local:ImageFileInfo">
    <Grid Height="300"
        Width="300"
        Margin="8">
        <Grid.RowDefinitions>
            <RowDefinition />
            <RowDefinition Height="Auto" />
        </Grid.RowDefinitions>

        <Image x:Name="ItemImage"
            Source="{x:Bind ImageUri}"
            Stretch="Uniform" />

        <StackPanel Orientation="Vertical"
                    Grid.Row="1">
            <TextBlock Text="{x:Bind ImageTitle}"
                    HorizontalAlignment="Center"
                    Style="{StaticResource SubtitleTextBlockStyle}" />
            <TextBlock Text="{x:Bind ImageFileType}"
                    HorizontalAlignment="Center"
                    Style="{StaticResource CaptionTextBlockStyle}" />
            <RatingControl Value="{x:Bind ImageRating}" IsReadOnly="True"/>
        </StackPanel>
    </Grid>

> </DataTemplate>

Reviewer 3: Same here XAML formatting

```

> ### Running the app

Reviewer 2: There are no instructions on how to run Mac one. But there is Linux, which is great.

Launch the `SimplePhotosUno.Windows` target. Observe that this WinUI app is very similar to the previous tutorial.

You can now build and run your app on any of the supported platforms. To do so, you can use the debug toolbar drop-down to select a target platform to deploy:

> * To run the **WebAssembly** (Wasm) head:

Reviewer 2: Consider using an image for each one of the platform-run , so they can see the final result.

[Insert picture here]

## Prerequisites

- [Visual Studio 2022 17.4 or later](https://visualstudio.microsoft.com/#vs-section)
- [Tools for Windows App SDK](../windows-app-sdk/set-up-your-development-environment.md)
- ASP.NET and web development workload (for WebAssembly development)
:::image type="content" source="../images/uno/uno-vs-install-web.png" alt-text="Web development workload in VS":::
- .NET Multi-platform App UI development installed (for iOS, Android, Mac Catalyst development).
:::image type="content" source="../images/uno/uno-vs-install-dotnet-mobile.png" alt-text="dotnet mobile workload in VS":::
- .NET desktop development installed (for Gtk, Wpf, and Linux Framebuffer development)
:::image type="content" source="../how-tos/images/hello-world/vs-install-dotnet.png" alt-text=".net desktop workload in VS":::

## Finalize your environment

> 1. Open a command-line prompt, Windows Terminal if you have it installed, or else Command Prompt or Windows Powershell from the Start Menu.

Reviewer 4: @alvinashcraft are we able to extract the "finalize your environment" steps (which are basically the steps for ensuring uno is setup) into a separate .md file that can be referenced both here and in [this doc](https://learn.microsoft.com/en-us/windows/apps/how-tos/uno-multiplatform) so that we can maintain these instructions in a single place?
Perhaps this should include the above pre-requisites too?

Reviewer 5: Hey Nick. Yes, you can use include files. Here's the guidance on them: https://learn.microsoft.com/en-us/contribute/markdown-reference#includes-syntax

Let me know if you have questions about using them. You can put the "Includes" folder in one spot and use a relative link to reference it from both articles. You should be able to search the repo for other uses to get an idea of how it's used in practice.

Reviewer 4: @lukeblevins can you give this a shot to share content between the two uno blog posts

                  ContainerContentChanging="ImageGridView_ContainerContentChanging">
        </GridView>
    </Grid>
</Page>
```

Recall that the desktop application also had a `MainWindow.xaml.cs` file that contained the code-behind for the view. In the Uno Platform project, the code-behind for the `MainPage` view is contained in the `MainPage.xaml.cs` file. Copy the methods and `Images` member from the `MainWindow.xaml.cs` file and paste them into the `MainPage.xaml.cs` file. The `MainPage.xaml.cs` file should look like this:

```csharp
using Microsoft.UI.Xaml.Controls;
using System.Collections.ObjectModel;
using Windows.Storage;
using Windows.Storage.Search;

namespace UnoSimplePhotos;

public sealed partial class MainPage : Page
{
    public ObservableCollection<ImageFileInfo> Images { get; } 
    = new ObservableCollection<ImageFileInfo>();

	public MainPage()
	{
		this.InitializeComponent();
        GetItemsAsync();
    }

    private void ImageGridView_ContainerContentChanging(ListViewBase sender,
        ContainerContentChangingEventArgs args)
    {
        if (args.InRecycleQueue)
        {
            var templateRoot = args.ItemContainer.ContentTemplateRoot as Grid;

>             var image = templateRoot.FindName("ItemImage") as Image;
>             image.Source = null;

Reviewer 4: ```suggestion
                var image = templateRoot?.FindName("ItemImage") as Image;
                if (image is not null)
                {
                    image.Source = null;
                }
```

        }

        if (args.Phase == 0)
        {
            args.RegisterUpdateCallback(ShowImage);
            args.Handled = true;
        }
    }

    private async void ShowImage(ListViewBase sender, ContainerContentChangingEventArgs args)
    {
        if (args.Phase == 1)
        {
            // It's phase 1, so show this item's image.
            var templateRoot = args.ItemContainer.ContentTemplateRoot as Grid;

>             var image = templateRoot.FindName("ItemImage") as Image;
>             var item = args.Item as ImageFileInfo;
>             image.Source = await item.GetImageThumbnailAsync();

Reviewer 4: ```suggestion
                var image = templateRoot?.FindName("ItemImage") as Image;
                var item = args.Item as ImageFileInfo;
                if (image is not null && item is not null)
                {
                    image.Source = await item.GetImageThumbnailAsync();
                }
```

        }
    }

    private async Task GetItemsAsync()
    {
        StorageFolder appInstalledFolder = Package.Current.InstalledLocation;
        StorageFolder picturesFolder = await appInstalledFolder.GetFolderAsync("Assets\\Samples");

        var result = picturesFolder.CreateFileQueryWithOptions(new QueryOptions());

        IReadOnlyList<StorageFile> imageFiles = await result.GetFilesAsync();
        foreach (StorageFile file in imageFiles)
        {
            Images.Add(await LoadImageInfoAsync(file));
        }

>         ImageGridView.ItemsSource = Images;

Reviewer 4: Remove this line - the Images property is data bound using the x:Bind expression on the GridView

    }

    public async static Task<ImageFileInfo> LoadImageInfoAsync(StorageFile file)
    {
        var properties = await file.Properties.GetImagePropertiesAsync();
        ImageFileInfo info = new(properties,
                                    file, file.DisplayName, file.DisplayType);

        return info;
    }
}
```

> The `MainPage.xaml.cs` file now contains methods from the desktop projects which we can modify in subsequent steps for multi-platform compatibility. We now need to copy the `ImageFileInfo` class from the desktop project and paste it into the `MainPage.xaml.cs` file. The `MainPage.xaml.cs` file should now look like this:

Reviewer 4: Should "MainPage.xaml.cs" be a new file called "ImageFileInfo.cs"?

```csharp
using Microsoft.UI.Xaml.Media.Imaging;
using System.ComponentModel;
using System.Runtime.CompilerServices;
using Windows.Storage;
using Windows.Storage.FileProperties;

> using Windows.Storage.Streams;

Reviewer 4: ```suggestion
using Windows.Storage.Streams;
using ThumbnailMode = Windows.Storage.FileProperties.ThumbnailMode;
```

> namespace UnoSimplePhotos;

Reviewer 4: Point out the namespace has been updated to UnoSimplePhotos

            await ImageFile.GetThumbnailAsync(ThumbnailMode.PicturesView);
        // Create a bitmap to be the image source.
        var bitmapImage = new BitmapImage();
        bitmapImage.SetSource(thumbnail);
        thumbnail.Dispose();

        return bitmapImage;
    }

    public string ImageName { get; }

    public string ImageFileType { get; }

    public string ImageDimensions => $"{ImageProperties.Width} x {ImageProperties.Height}";

    public string ImageTitle
    {
        get => string.IsNullOrEmpty(ImageProperties.Title) ? ImageName : ImageProperties.Title;
        set
        {
            if (ImageProperties.Title != value)
            {
                ImageProperties.Title = value;
                _ = ImageProperties.SavePropertiesAsync();
                OnPropertyChanged();
            }
        }
    }

    public int ImageRating
    {
        get => (int)ImageProperties.Rating;
        set
        {
            if (ImageProperties.Rating != value)
            {
                ImageProperties.Rating = (uint)value;
                _ = ImageProperties.SavePropertiesAsync();
                OnPropertyChanged();
            }
        }
    }

>     public event PropertyChangedEventHandler PropertyChanged;

Reviewer 4: ```suggestion
    public event PropertyChangedEventHandler? PropertyChanged;
```
Make PropertyChangedEventHandler nullable to avoid build warning

>     protected void OnPropertyChanged([CallerMemberName] string propertyName = null) =>

Reviewer 4: ```suggestion
    protected void OnPropertyChanged([CallerMemberName] string? propertyName = null) =>
```

        PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(propertyName));
}
```

> This class will serve as a model to represent the image files in the `GridView`. In the next sections, we will make a set of changes to these copied files to make them compatible in a multi-platform context.

Reviewer 4: Should the solution build and run at this point?
So far you've given a lot of instructions, it would be good for the user to know whether they've done things correctly by either building and/or running the application (even if it doesn't render correctly)

> ### Add image assets to the project

Reviewer 4: This section should be before the "In the next sections, we will make a set of changes ... to make them compatible in a multi-platform context"

> In the `SimplePhotosUno` project, create a new folder named `Assets` and copy the JPEG image files to a `Samples` subfolder. The `Assets` folder structure should now look like this:

Reviewer 4: Whilst the files are technically JPEG, I would change this to JPG files, since that's the actual file extension.

Also, are the uno platform docs that provide guidance on adding assets? do assets always need to be in the Assets folder?

[Insert image here]

### Making the code multi-platform capable

In the desktop project from the previous tutorial, the `MainPage.xaml.cs` file contains a `GetItemsAsync` method that enumerates items from a `StorageFolder` representing the installed package location. Because that location is not available on certain platforms such as WebAssembly, we will need to make changes to this method to make it compatible with all platforms. We will accordingly make some changes to the `ImageFileInfo` class to ensure compatibility.

First, we will make changes to the `GetItemsAsync` method. Replace the `GetItemsAsync` method in the `MainPage.xaml.cs` file with the following code:

```csharp

> private async void GetItems()

Reviewer 4: This should be GetItemsAsync and we should return Task so that it can be awaited

                        <RatingControl Value="{x:Bind ImageRating}" 
                                       IsReadOnly="True"/>
                    </StackPanel>
                </Grid>
            </DataTemplate>

            <Style x:Key="ImageGridView_ItemContainerStyle"
                   TargetType="GridViewItem">
                <Setter Property="Background" Value="Gray"/>
                <Setter Property="Margin" Value="8"/>
            </Style>

            <ItemsPanelTemplate x:Key="ImageGridView_ItemsPanelTemplate">
                    <ItemsWrapGrid Orientation="Horizontal"
                                   HorizontalAlignment="Center"/>
                </ItemsPanelTemplate>
        </Grid.Resources>

        <GridView x:Name="ImageGridView"
                  ItemsSource="{x:Bind Images}"
                  ItemTemplate="{StaticResource ImageGridView_ItemTemplate}"
                  ItemContainerStyle="{StaticResource ImageGridView_ItemContainerStyle}"
                  ItemsPanel="{StaticResource ImageGridView_ItemsPanelTemplate}"
                  ContainerContentChanging="ImageGridView_ContainerContentChanging" />
    </Grid>
</Window>
```

> Uno Platform's multi-platform implmentation of the controls found in the `Window` element, such as `GridView`, `Image`, and `RatingControl`, ensure that the view itself will work on all supported platforms with only a trivial amount of effort. Copy the contents of this `Window` and paste them into the `Page` element of the `MainPage.xaml` file in the **UnoSimplePhotos** Uno Platform project. The `MainPage` view XAML should look like this:

Reviewer: ```suggestion
Uno Platform's multi-platform implementation of the controls found in the `Window` element, such as `GridView`, `Image`, and `RatingControl`, ensure that the view itself will work on all supported platforms with only a trivial amount of effort. Copy the contents of this `Window` and paste them into the `Page` element of the `MainPage.xaml` file in the **UnoSimplePhotos** Uno Platform project. The `MainPage` view XAML should look like this:
```

```xml

> <Page
>     x:Class="UnoSimplePhotos.MainPage"
>     xmlns="http://schemas.microsoft.com/winfx/2006/xaml/presentation"
>     xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
>     xmlns:local="using:UnoSimplePhotos"
>     xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
>     xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
>     mc:Ignorable="d">

Reviewer 3: ```suggestion
<Page x:Class="UnoSimplePhotos.MainPage"
      xmlns="http://schemas.microsoft.com/winfx/2006/xaml/presentation"
      xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
      xmlns:local="using:UnoSimplePhotos"
      xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
      xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
      mc:Ignorable="d">
```

                <Setter Property="Background" 
                        Value="Gray"/>
                <Setter Property="Margin" 
                        Value="8"/>
            </Style>

            <ItemsPanelTemplate x:Key="ImageGridView_ItemsPanelTemplate">
                <ItemsWrapGrid Orientation="Horizontal"
                               HorizontalAlignment="Center"/>
            </ItemsPanelTemplate>
        </Grid.Resources>

        <GridView x:Name="ImageGridView"
                  ItemsSource="{x:Bind Images}"
                  ItemTemplate="{StaticResource ImageGridView_ItemTemplate}"
                  ItemContainerStyle="{StaticResource ImageGridView_ItemContainerStyle}"
                  ItemsPanel="{StaticResource ImageGridView_ItemsPanelTemplate}"
                  ContainerContentChanging="ImageGridView_ContainerContentChanging">
        </GridView>
    </Grid>
</Page>
```

Recall that the desktop solution also had a `MainWindow.xaml.cs` file that contained code-behind which corresponds to the view. In the Uno Platform project, the code-behind for the `MainPage` view we've copied into is contained in the `MainPage.xaml.cs` file. 

> To bring this codebehind multiplatform, we should first move the following into the `MainPage.xaml.cs` file:

Reviewer: ```suggestion
To bring this code-behind multi-platform, we should first move the following into the `MainPage.xaml.cs` file:
```

"code-behind" * "multi-platform" are hyphenated throughout the rest of the document

> - `Images` property: Provides the `GridView` with an observable collection of image files
> 
> - Contents of constructor: Calls `GetItemsAsync()` to populate the `Images` collection with items representing image files
> 
> - Remove the manual modification of the `ImageGridView` control's `ItemsSource` property
> 
> - `ImageGridView_ContainerContentChanging` method: Used as part of a [strategy](https://learn.microsoft.com/windows/uwp/debug-test-perf/optimize-gridview-and-listview#update-listview-and-gridview-items-progressively) to progressively load `GridView` items as they are scrolled into view
> 
> - `ShowImage` method: Loads the image files into the `GridView`
> 
> - `GetItemsAsync` method: Gets the image asset files from the `Samples` folder
> 
> - `LoadImageInfoAsync` method: Constructs a `ImageFileInfo` object from a created `StorageFile`
> 
> After moving everything over, `MainPage.xaml.cs` should now look like this:

Reviewer 3: ```suggestion
- `Images` property: Provides the `GridView` with an observable collection of image files

- Contents of the constructor: Calls `GetItemsAsync()` to populate the `Images` collection with items representing image files

- Remove the manual modification of the `ImageGridView` control's `ItemsSource` property

- `ImageGridView_ContainerContentChanging` method: Used as part of a [strategy](https://learn.microsoft.com/windows/uwp/debug-test-perf/optimize-gridview-and-listview#update-listview-and-gridview-items-progressively) to progressively load `GridView` items as they are scrolled into view

- `ShowImage` method: Loads the image files into the `GridView`

- `GetItemsAsync` method: Gets the image asset files from the `Samples` folder

- `LoadImageInfoAsync` method: Constructs an `ImageFileInfo` object from a created `StorageFile`

After moving everything over, `MainPage.xaml.cs` should now look like this:
```

        }
#else
    ShowImage(sender, args);
#endif
}
```

The specialized callback is now only registered using `ContainerContentChangingEventArgs.RegisterUpdateCallback()` if the platform is Windows. Otherwise, the `ShowImage` method is called directly. We will also need to make changes to the `ShowImage` method to work alongside the changes made to the `ImageGridView_ContainerContentChanging` method. Replace the `ShowImage` method in the `MainPage.xaml.cs` file with the following code:

```csharp
private async void ShowImage(ListViewBase sender, ContainerContentChangingEventArgs args)
{
    if (
#if WINDOWS
            args.Phase == 1
#else
        true
#endif
        )
    {

        // It's phase 1, so show this item's image.
        var templateRoot = args.ItemContainer.ContentTemplateRoot as Grid;
        var image = templateRoot?.FindName("ItemImage") as Image;
        var item = args.Item as ImageFileInfo;
#if WINDOWS
        if (image is not null && item is not null)
        {
            image.Source = await item.GetImageThumbnailAsync();
        }
#else
        if (item is not null)
        {
            await item.GetImageSourceAsync();
        }
#endif
    }
}
```

> Again, preprocessor directives ensure that the `ContainerContentChangingEventArgs.Phase` property is only used where it is supported. We make use of the previously unused `GetImageSourceAsync()` method to load the image files into the `GridView` on platforms other than Windows. At this point, we will accommodate the changes made above by editing to the `ImageFileInfo` class. 

Reviewer 3: ```suggestion
Again, preprocessor directives ensure that the `ContainerContentChangingEventArgs.Phase` property is only used where it is supported. We make use of the previously unused `GetImageSourceAsync()` method to load the image files into the `GridView` on platforms other than Windows. At this point, we will accommodate the changes made above by editing the `ImageFileInfo` class. 
```

### Creating a separate code path for other platforms

Update `ImageFileInfo.cs` to include a new property called `ImageSource` that will be used to load the image file.

```csharp
public BitmapImage? ImageSource { get; private set; }
```

> Because platforms like the Web do not support advanced image file properties that are readily available on Windows, we should add a constructor overload which does not require an `ImageProperties` typed parameter. Add a new overload under the existing one using the following code:

Reviewer 3: ```suggestion
Because platforms like the Web do not support advanced image file properties that are readily available on Windows, we should add a constructor overload that does not require an `ImageProperties` typed parameter. Add a new overload under the existing one using the following code:
```

```csharp
public ImageFileInfo(StorageFile imageFile,
    string name,
    string type)
{
    ImageName = name;
    ImageFileType = type;
    ImageFile = imageFile;
}
```

This constructor overload will be used to construct an `ImageFileInfo` object on platforms other than Windows. Since we did this, it makes sense to make the `ImageProperties` property nullable. Update the `ImageProperties` property to be nullable using the following code:

```csharp
public ImageProperties? ImageProperties { get; }
```

Update the `GetImageSourceAsync` method to use the `ImageSource` property instead of only returning a `BitmapImage` object. Replace the `GetImageSourceAsync` method in the `ImageFileInfo.cs` file with the following code:

```csharp
public async Task<BitmapImage> GetImageSourceAsync()
{
    using IRandomAccessStream fileStream = await ImageFile.OpenReadAsync();

    // Create a bitmap to be the image source.
    BitmapImage bitmapImage = new();
    bitmapImage.SetSource(fileStream);

    ImageSource = bitmapImage;
    PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(nameof(ImageSource)));

    return bitmapImage;
}
```

> To prevent getting the value of `ImageProperties` when its null, we will need to make the following changes:

Reviewer 3: ```suggestion
To prevent getting the value of `ImageProperties` when it's null, we will need to make the following changes:
```

```

The `MainPage.xaml` file should now look like this:

```xml
<Page x:Class="UnoSimplePhotos.MainPage"
      xmlns="http://schemas.microsoft.com/winfx/2006/xaml/presentation"
      xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
      xmlns:local="using:UnoSimplePhotos"
      xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
      xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
      xmlns:win="http://schemas.microsoft.com/winfx/2006/xaml/presentation"
      mc:Ignorable="d"
      Background="{ThemeResource ApplicationPageBackgroundThemeBrush}">

	<Grid>
        <Grid.Resources>
            <DataTemplate x:Key="ImageGridView_ItemTemplate"
                          x:DataType="local:ImageFileInfo">
                <Grid Height="300"
                      Width="300"
                      Margin="8">
                    <Grid.RowDefinitions>
                        <RowDefinition />
                        <RowDefinition Height="Auto" />
                    </Grid.RowDefinitions>

                    <Image x:Name="ItemImage"
                           Source="{x:Bind ImageSource}"
                           Stretch="Uniform" />

>                     <StackPanel Orientation="Vertical"
> 								Grid.Row="1">

Reviewer 3: ```suggestion
                    <StackPanel Orientation="Vertical"
				Grid.Row="1">
```
Format properly the rest of the XAML below also

                        Value="Gray" />
                <Setter Property="Margin" 
                        Value="8"/>
            </Style>

            <ItemsPanelTemplate x:Key="ImageGridView_ItemsPanelTemplate">
                <ItemsWrapGrid Orientation="Horizontal"
                               HorizontalAlignment="Center"/>
            </ItemsPanelTemplate>
        </Grid.Resources>

        <GridView x:Name="ImageGridView"
                  ItemsSource="{x:Bind Images, Mode=OneWay}"
                  win:ItemsPanel="{StaticResource ImageGridView_ItemsPanelTemplate}"
                  ContainerContentChanging="ImageGridView_ContainerContentChanging"
                  ItemContainerStyle="{StaticResource ImageGridView_ItemContainerStyle}"
                  ItemTemplate="{StaticResource ImageGridView_ItemTemplate}" />
    </Grid>
</Page>
```

### Running the app

Launch the `UnoSimplePhotos.Windows` target. Observe that this WinUI app is very similar to the previous tutorial.

You can now build and run your app on any of the supported platforms. To do so, you can use the debug toolbar drop-down to select a target platform to deploy:

* To run the **WebAssembly** (Wasm) head:

>     - Right-click on the `UnoSimplePhotos.Wasm` project, select **Set as startup project**
>     - Press the `UnoSimplePhotos.Wasm` button to deploy the app
>     - If desired, you can add and use the `UnoSimplePhotos.Server` project as an alternative
> * To debug for **iOS**:
>     - Right-click on the `UnoSimplePhotos.Mobile` project, select **Set as startup project**
>     - In the debug toolbar drop-down, select an active iOS device or the simulator. You'll need to be paired with a Mac for this to work.
> 
>       :::image type="content" source="../images/uno/uno-mobile-debug.png" alt-text="Visual Studio dropdown to select a target framework to deploy":::
> 
> * To debug for **Mac Catalyst**: 
>     - Right-click on the `UnoSimplePhotos.Mobile` project, select **Set as startup project**
>     - In the debug toolbar drop-down, select a remote macOS device. You'll need to be paired with one for this to work.
> * To debug the **Android** platform:
>     - Right-click on the `UnoSimplePhotos.Mobile` project, select **Set as startup project**
>     - In the debug toolbar drop-down, select either an active Android device or the emulator
>         - Select an active device in the "Device" sub-menu
> * To debug on **Linux** with **Skia GTK**:
>     - Right-click on the `UnoSimplePhotos.Skia.Gtk` project, and select **Set as startup project**
>     - Press the `UnoSimplePhotos.Skia.Gtk` button to deploy the app

Reviewer 3: ```suggestion
    - Right-click on the `UnoSimplePhotos.Wasm` project, select **Set as startup project**
    - Press the `UnoSimplePhotos.Wasm` button to deploy the app
    - If desired, you can add and use the `UnoSimplePhotos.Server` project as an alternative
* To debug for **iOS**:
    - Right-click on the `UnoSimplePhotos.Mobile` project, and select **Set as startup project**
    - In the debug toolbar drop-down, select an active iOS device or the simulator. You'll need to be paired with a Mac for this to work.

      :::image type="content" source="../images/uno/uno-mobile-debug.png" alt-text="Visual Studio dropdown to select a target framework to deploy":::

* To debug for **Mac Catalyst**: 
    - Right-click on the `UnoSimplePhotos.Mobile` project, and select **Set as startup project**
    - In the debug toolbar drop-down, select a remote macOS device. You'll need to be paired with one for this to work.
* To debug the **Android** platform:
    - Right-click on the `UnoSimplePhotos.Mobile` project, and select **Set as startup project**
    - In the debug toolbar drop-down, select either an active Android device or the emulator
        - Select an active device in the "Device" sub-menu
* To debug on **Linux** with **Skia GTK**:
    - Right-click on the `UnoSimplePhotos.Skia.Gtk` project, and select **Set as startup project**
    - Press the `UnoSimplePhotos.Skia.Gtk` button to deploy the app
```
